### PR TITLE
Tree: test cursor paths

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Adds path comparisons utilities and uses them to check that paths reported from cursors are correct.

### CODE REVIEW

File: packages/dds/tree/src/test/cursorTestSuite.ts [hunk 1]
```diff
@@ -489,14 +493,28 @@ function checkTraversal(cursor: ITreeCursor) {
         assert(!fieldLengths.has(key), "no duplicate keys");
         fieldLengths.set(cursor.getFieldKey(), expectedFieldLength);
         assert(expectedFieldLength > 0, "only non empty fields should show up in field iteration");
+        assert(compareFieldUpPaths(cursor.getFieldPath(), { field: key, parent: path }));
+
+        // Check that iterating nodes withing field works as expected.
```
Comment (jenn-le) [inline packages/dds/tree/src/test/cursorTestSuite.ts:498]:
```suggestion
        // Check that iterating nodes within fields works as expected.
```

File: packages/dds/tree/src/test/cursorTestSuite.ts [hunk 2]
```diff
@@ -505,7 +523,7 @@ function checkTraversal(cursor: ITreeCursor) {
             "Did not traverse expected number of children",
         );
 
-        // Cheek field access by index
+        // Cheek node access by index
```
Comment (jenn-le) [inline packages/dds/tree/src/test/cursorTestSuite.ts:526]:
```suggestion
        // Check node access by index
```

File: packages/dds/tree/src/test/cursorTestSuite.ts [hunk 3]
```diff
@@ -489,14 +493,28 @@ function checkTraversal(cursor: ITreeCursor) {
...
         for (let inNode = cursor.firstNode(); inNode; inNode = cursor.nextNode()) {
             assert(cursor.chunkStart <= actualChildNodesTraversed);
             assert(cursor.chunkLength > actualChildNodesTraversed - cursor.chunkStart);
             assert(cursor.chunkLength + cursor.chunkStart <= expectedFieldLength);
+
+            // Make sure down+up navigation gets back to where it started.
+            // Testing this explicitly here before recursing makes debugging issues with this easier.
+            assert.equal(cursor.fieldIndex, actualChildNodesTraversed);
+            cursor.enterField(EmptyKey);
+            cursor.exitField();
```
Comment (jenn-le) [inline packages/dds/tree/src/test/cursorTestSuite.ts:509]: What does this test? IIRC, doesn't this just push EmptyKey to the navigation stack and then pop it without doing any sort of validation?

File: packages/dds/tree/src/test/cursorTestSuite.ts [hunk 4]
```diff
@@ -546,6 +564,19 @@ function checkTraversal(cursor: ITreeCursor) {
         // skipPendingFields should have no effect since not pending
         assert(cursor.skipPendingFields());
         assert.equal(cursor.getFieldKey(), key);
+
+        // Recursively validate.
+        actualChildNodesTraversed = 0;
+        for (let inNode = cursor.firstNode(); inNode; inNode = cursor.nextNode()) {
```
Comment (jenn-le) [inline packages/dds/tree/src/test/cursorTestSuite.ts:570]: What's the purpose of testing this inside a separate loop?

**Review by jenn-le — APPROVED** [review]

File: packages/dds/tree/src/test/cursorTestSuite.ts (hunk 1 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/test/cursorTestSuite.ts:498]:
```suggestion
        // Check that iterating nodes of this field works as expected.
```

File: packages/dds/tree/src/test/cursorTestSuite.ts (hunk 4 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/test/cursorTestSuite.ts:570]: As noted in the comment above, it's easier to debug non-recursive validation, so we do the non-recursive validation first, making sure as much of the cursor is working as practical before we do the recursive check: if basic navigation is broken, the recursive check can produce really confusing errors so I prefer more specific errors I can know were not caused by the recursion and ca easily step through with the debugger.

This means validating both iterating and indexing done on the above loops before recursing.

**Review by jenn-le — APPROVED** [review]

**Pull request merged**